File: screenshots/screenshot_test.dart
import 'package:hive/hive.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:provider/provider.dart';
import 'package:timety/main.dart' as app;
import 'package:timety/data/focus/focus_models.dart';
import 'package:timety/data/focus/focus_repository_hive.dart';
import 'package:timety/data/habit/habit_models.dart';
import 'package:timety/data/habit/habit_repository_hive.dart';
import 'package:timety/data/task/task.dart';
import 'package:timety/data/task/task_repository_hive.dart';
import 'package:timety/providers/focus_provider.dart';
import 'package:timety/providers/habit_provider.dart';
import 'package:timety/providers/settings_provider.dart';
import 'package:timety/providers/task_provider.dart';
import 'package:timety/providers/user_provider.dart';
import 'package:timety/theme/app_theme.dart';
import 'package:timety/utils/stats/xp_calculator.dart';

// How to execute:
// 1. Start Emulator
// 2. Execute: flutter drive --driver=screenshots/test_driver.dart --target=screenshots/screenshot_test.dart

// --- Pump Loop ---
// Fast-forwards the clock in 50ms increments to allow animations/charts to render
Future<void> _pumpDuration(WidgetTester tester, Duration duration) async {
  final int frameCount = (duration.inMilliseconds / 50).round();
  for (int i = 0; i < frameCount; i++) {
    await tester.pump(const Duration(milliseconds: 50));
  }
}

Future<void> _waitForCondition(
  WidgetTester tester,
  bool Function() condition, {
  Duration timeout = const Duration(seconds: 30),
}) async {
  final stopwatch = Stopwatch()..start();

  while (!condition()) {
    if (stopwatch.elapsed > timeout) {
      throw Exception('Timed out while waiting for the screenshot app state.');
    }
    await tester.pump(const Duration(milliseconds: 100));
  }
}

Future<void> _tapBottomNavItem(WidgetTester tester, String label) async {
  final finder = find.descendant(
    of: find.byType(BottomNavigationBar),
    matching: find.text(label),
  );

  await tester.tap(finder.first);
  // Replaced static pump with the pump loop (1.5 seconds is perfect for page transitions)
  await _pumpDuration(tester, const Duration(milliseconds: 1500));
}

Future<void> _tapIconButton(
  WidgetTester tester,
  IconData icon, {
  int index = 0,
}) async {
  final finder = find.byIcon(icon);
  if (finder.evaluate().isEmpty) {
    throw Exception('Icon $icon not found');
  }

  await tester.tap(finder.at(index));
  // Replaced static pump with the pump loop
  await _pumpDuration(tester, const Duration(milliseconds: 1500));
}

Future<void> _tapText(WidgetTester tester, String text) async {
  final finder = find.text(text);
  if (finder.evaluate().isEmpty) {
    throw Exception('Text "$text" not found');
  }

  await tester.tap(finder.first);
  // Replaced static pump with the pump loop
  await _pumpDuration(tester, const Duration(milliseconds: 1500));
}

Future<void> _resetPersistentData() async {
  final tasksBox = await Hive.openBox<Task>(HiveTaskRepository.boxName);
  await tasksBox.clear();

  final habitsBox = await Hive.openBox<Habit>(HiveHabitRepository.boxName);
  await habitsBox.clear();

  final sessionsBox = await Hive.openBox<FocusSession>(
    HiveFocusRepository.sessionBoxName,
  );
  await sessionsBox.clear();

  final tagsBox = await Hive.openBox<FocusTag>(HiveFocusRepository.tagBoxName);
  await tagsBox.clear();
}

Future<void> _seedMockData(BuildContext context) async {
  final settings = context.read<SettingsProvider>();
  final taskProvider = context.read<TaskProvider>();
  final habitProvider = context.read<HabitProvider>();
  final focusProvider = context.read<FocusProvider>();
  final userProvider = context.read<UserProvider>();

  await userProvider.updateName("Bobert");
  await userProvider.addXp(-userProvider.totalXp);
  settings.setThemeMode(ThemeMode.light);
  settings.setDailyGoal(120);

  taskProvider.tasks.clear();
  habitProvider.habits.clear();
  focusProvider.history.clear();
  focusProvider.tags.clear();

  final today = DateTime.now();
  final dayStart = DateTime(today.year, today.month, today.day);

  await habitProvider.saveHabit(
    Habit(
      id: 'habit_water',
      name: 'Drink Water',
      frequency: HabitFrequency.daily,
      stackName: 'Morning Routine',
      stackOrder: 1,
      colorValue: AppTheme.habitColor.toARGB32(),
    ),
  );

  await habitProvider.saveHabit(
    Habit(
      id: 'habit_meditate',
      name: 'Meditation',
      frequency: HabitFrequency.daily,
      stackName: 'Morning Routine',
      stackOrder: 2,
      completions: [dayStart.add(const Duration(hours: 7, minutes: 15))],
      colorValue: AppTheme.habitColor.toARGB32(),
    ),
  );

  await habitProvider.saveHabit(
    Habit(
      id: 'habit_read',
      name: 'Read 20 Pages',
      frequency: HabitFrequency.weeklyExact,
      targetWeekdays: [today.weekday],
      notes: 'Keep the momentum going before lunch.',
      colorValue: AppTheme.habitColor.toARGB32(),
    ),
  );

  await habitProvider.saveHabit(
    Habit(
      id: 'habit_workout',
      name: 'Workout',
      frequency: HabitFrequency.weeklyFlexible,
      targetDaysPerWeek: 3,
      completions: [
        dayStart
            .subtract(const Duration(days: 3))
            .add(const Duration(hours: 18)),
        dayStart
            .subtract(const Duration(days: 2))
            .add(const Duration(hours: 18)),
        dayStart
            .subtract(const Duration(days: 1))
            .add(const Duration(hours: 18)),
      ],
      colorValue: AppTheme.habitColor.toARGB32(),
    ),
  );

  await taskProvider.addTask(
    Task(
      id: 'task_proposal',
      title: 'Finish project proposal',
      description: 'Polish the scope, milestones, and budget notes.',
      dueDate: dayStart.subtract(const Duration(hours: 3)),
      location: 'Home office',
      priority: Priority.high,
      size: Size.large,
      category: 'Work',
      createdAt: dayStart.subtract(const Duration(days: 2)),
      subtasks: [
        Subtask(id: 'task_proposal_1', title: 'Review outline'),
        Subtask(id: 'task_proposal_2', title: 'Check budget table'),
      ],
    ),
  );

  await taskProvider.addTask(
    Task(
      id: 'task_client',
      title: 'Reply to client email',
      description:
          'Answer the questions from yesterday'
          's review.',
      dueDate: dayStart.add(const Duration(hours: 23)),
      location: 'Inbox',
      priority: Priority.veryHigh,
      size: Size.small,
      category: 'Communication',
      createdAt: dayStart.subtract(const Duration(days: 1)),
      reminders: [dayStart.add(const Duration(hours: 9, minutes: 30))],
    ),
  );

  await taskProvider.addTask(
    Task(
      id: 'task_retro',
      title: 'Prepare sprint retro',
      description: 'Collect highlights and blockers for Friday.',
      dueDate: dayStart.add(const Duration(days: 2)),
      location: 'Meeting room',
      category: 'Work',
      createdAt: dayStart,
    ),
  );

  await taskProvider.addTask(
    Task(
      id: 'task_backup',
      title: 'Back up phone photos',
      description: 'Archive the last trip and mark it complete.',
      dueDate: dayStart.subtract(const Duration(days: 1)),
      location: 'Laptop',
      priority: Priority.low,
      size: Size.small,
      category: 'Personal',
      isCompleted: true,
      completedAt: dayStart.subtract(const Duration(hours: 20)),
      createdAt: dayStart.subtract(const Duration(days: 4)),
    ),
  );

  await focusProvider.createTag('Deep Work', Colors.indigo);
  final deepWorkTag = focusProvider.selectedTag!;

  final flexibleMode = focusProvider.modes.firstWhere(
    (mode) => mode.id == 'system_flexible',
  );
  focusProvider.setActiveMode(flexibleMode);
  focusProvider.setFlexibleDuration(50);

  await focusProvider.logPastSession(
    mode: flexibleMode,
    startTime: dayStart.add(const Duration(hours: 8, minutes: 15)),
    endTime: dayStart.add(const Duration(hours: 9)),
    tag: focusProvider.selectedTag,
  );

  await focusProvider.logPastSession(
    mode: flexibleMode,
    startTime: dayStart.add(const Duration(hours: 13, minutes: 30)),
    endTime: dayStart.add(const Duration(hours: 14, minutes: 10)),
    tag: deepWorkTag,
  );

  await focusProvider.logPastSession(
    mode: FocusMode.classicPomodoro(),
    startTime: dayStart
        .subtract(const Duration(days: 1))
        .add(const Duration(hours: 16)),
    endTime: dayStart
        .subtract(const Duration(days: 1))
        .add(const Duration(hours: 16, minutes: 30)),
    tag: deepWorkTag,
  );

  final completedTasks = taskProvider.tasks.where((t) => t.isCompleted).length;
  final totalHabitCompletions = habitProvider.habits.fold(
    0,
    (sum, h) => sum + h.completions.length,
  );
  final totalFocusMinutes = focusProvider.history.fold(
    0,
    (sum, s) => sum + (s.totalSecondsFocused ~/ 60),
  );

  final seededXp =
      (completedTasks * ExperienceEngine.xpPerTask) +
      (totalHabitCompletions * ExperienceEngine.xpPerHabit) +
      (totalFocusMinutes * ExperienceEngine.xpPerFocusMin);

  await userProvider.addXp(seededXp);
}

void main() {
  final binding = IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets('Generate App Screenshots', (WidgetTester tester) async {
    // 1. Start the app
    app.main();

    await _waitForCondition(tester, () {
      if (find.byType(Scaffold).evaluate().isEmpty) {
        return false;
      }

      final focusProvider = Provider.of<FocusProvider>(
        tester.element(find.byType(Scaffold).first),
        listen: false,
      );
      return focusProvider.activeMode != null;
    });

    // Settle any remaining initial entry animations
    await _pumpDuration(tester, const Duration(milliseconds: 1000));

    await _resetPersistentData();

    // 2. Inject Dummy Data - capture context after async operations
    final BuildContext context = tester.element(find.byType(Scaffold).first);
    if (context.mounted) {
      await _seedMockData(context);
    }

    // Wait for the UI to visually update with the new data
    await binding.convertFlutterSurfaceToImage();
    await _pumpDuration(tester, const Duration(milliseconds: 1000));

    // 3. Capture the major app screens.
    await binding.takeScreenshot('01_home_screen');

    await _tapBottomNavItem(tester, 'Focus');
    await binding.takeScreenshot('02_focus_screen');

    // Navigate to Focus Modes from Focus screen by tapping the mode name
    await _tapText(tester, 'FLEXIBLE');
    await binding.takeScreenshot('06_focus_modes_screen');

    // Go back to Focus tab
    await tester.pageBack();
    await _pumpDuration(tester, const Duration(milliseconds: 1500));

    // Navigate to Profile screen
    await _tapBottomNavItem(tester, 'Profile');

    // Navigate to Stats from Profile screen (bar chart icon)
    await _tapIconButton(tester, Icons.bar_chart);
    await binding.takeScreenshot('07_stats_screen');

    // Go back to Profile tab
    await tester.pageBack();
    await _pumpDuration(tester, const Duration(milliseconds: 1500));

    await _tapBottomNavItem(tester, 'Tasks');
    await binding.takeScreenshot('03_tasks_screen');

    // Navigate to Calendar from Tasks
    await _tapIconButton(tester, Icons.calendar_today);
    await binding.takeScreenshot('08_calendar_screen');

    // Go back to Tasks
    await tester.pageBack();
    await _pumpDuration(tester, const Duration(milliseconds: 1500));

    // Tap on the first task to see detail view (the overdue proposal task)
    await _tapText(tester, 'Finish project proposal');
    await binding.takeScreenshot('09_task_detail_screen');

    // Go back to Tasks
    await tester.pageBack();
    await _pumpDuration(tester, const Duration(milliseconds: 1500));

    await _tapBottomNavItem(tester, 'Habits');
    await binding.takeScreenshot('04_habits_screen');

    // Tap on the first habit to see detail view
    await _tapText(tester, 'Drink Water');
    await binding.takeScreenshot('10_habit_detail_screen');

    // Go back to Habits
    await tester.pageBack();
    await _pumpDuration(tester, const Duration(milliseconds: 1500));

    await _tapBottomNavItem(tester, 'Profile');
    await binding.takeScreenshot('05_profile_screen');
  });
}

File: app/src/androidTest/java/io/github/benji377/timety/ScreenshotTest.kt
package io.github.benji377.timety

import android.graphics.Bitmap
import androidx.compose.ui.graphics.asAndroidBitmap
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.test.captureToImage
import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onAllNodesWithText
import androidx.compose.ui.test.onNodeWithContentDescription
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.onRoot
import androidx.compose.ui.test.performClick
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.GrantPermissionRule
import io.github.benji377.timety.data.model.focus.FocusModeEntity
import io.github.benji377.timety.data.model.focus.FocusModeType
import io.github.benji377.timety.data.model.focus.FocusSessionEntity
import io.github.benji377.timety.data.model.focus.FocusTagEntity
import io.github.benji377.timety.data.model.focus.FocusTargetType
import io.github.benji377.timety.data.model.focus.PhaseType
import io.github.benji377.timety.data.model.focus.SessionPhaseEntity
import io.github.benji377.timety.data.model.habit.HabitCompletionEntity
import io.github.benji377.timety.data.model.habit.HabitEntity
import io.github.benji377.timety.data.model.habit.HabitFrequency
import io.github.benji377.timety.data.model.task.Priority
import io.github.benji377.timety.data.model.task.SubtaskEntity
import io.github.benji377.timety.data.model.task.TaskEntity
import io.github.benji377.timety.data.model.task.TaskSize
import io.github.benji377.timety.data.model.user.UserProfileEntity
import io.github.benji377.timety.data.repository.ThemeMode
import io.github.benji377.timety.di.AppContainer
import io.github.benji377.timety.ui.theme.HabitColor
import io.github.benji377.timety.util.stats.ExperienceEngine
import kotlinx.coroutines.runBlocking
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.runner.RunWith
import java.io.File
import java.io.FileOutputStream
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId


@RunWith(AndroidJUnit4::class)
class ScreenshotTest {

    // Granted before the activity launches; otherwise MainScreen's permission prompt covers the
    // freshly launched activity (the test APK is reinstalled each run, revoking prior grants).
    val permissionRule: GrantPermissionRule =
        GrantPermissionRule.grant(android.Manifest.permission.POST_NOTIFICATIONS)

    val composeTestRule = createAndroidComposeRule<MainActivity>()

    @get:Rule
    val ruleChain: RuleChain = RuleChain.outerRule(permissionRule).around(composeTestRule)

    private val targetContext get() = InstrumentationRegistry.getInstrumentation().targetContext

    private fun str(resId: Int): String = targetContext.getString(resId)

    @Test
    fun generateScreenshots() {
        val container = (targetContext.applicationContext as TimetyApplication).container
        runBlocking {
            resetPersistentData(container)
            seedMockData(container)
        }

        // Wait until the seeded data has flowed into the Home screen.
        composeTestRule.waitUntil(timeoutMillis = 15_000) {
            composeTestRule.onAllNodesWithText("Reply to client email", substring = true)
                .fetchSemanticsNodes().isNotEmpty()
        }
        composeTestRule.waitForIdle()

        // 1. Home
        takeScreenshot("01_home_screen")

        // Statistics (bar chart icon in the Home app bar)
        composeTestRule.onNodeWithContentDescription(str(R.string.commonTooltipStats))
            .performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("07_stats_screen")
        pressBack()

        // Calendar (calendar icon in the Home app bar)
        composeTestRule.onNodeWithContentDescription(str(R.string.commonTooltipCalendar))
            .performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("08_calendar_screen")
        pressBack()

        // 2. Focus tab
        composeTestRule.onNodeWithText(str(R.string.navigationFocus)).performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("02_focus_screen")

        // Focus modes (manage-modes icon in the Focus app bar)
        composeTestRule.onNodeWithContentDescription(str(R.string.focusModesTitle)).performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("06_focus_modes_screen")
        pressBack()

        // 3. Tasks tab
        composeTestRule.onNodeWithText(str(R.string.navigationTasks)).performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("03_tasks_screen")

        // Task detail (the overdue proposal task)
        composeTestRule.onNodeWithText("Finish project proposal").performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("09_task_detail_screen")
        pressBack()

        // 4. Habits tab
        composeTestRule.onNodeWithText(str(R.string.navigationHabits)).performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("04_habits_screen")

        // Habit detail (first habit of the morning stack)
        composeTestRule.onNodeWithText("MORNING ROUTINE").performClick()
        composeTestRule.waitForIdle()
        composeTestRule.onNodeWithText("Drink Water").performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("10_habit_detail_screen")
        pressBack()

        // 5. Profile tab
        composeTestRule.onNodeWithText(str(R.string.navigationProfile)).performClick()
        composeTestRule.waitForIdle()
        takeScreenshot("05_profile_screen")
    }

    private fun pressBack() {
        composeTestRule.activityRule.scenario.onActivity {
            it.onBackPressedDispatcher.onBackPressed()
        }
        composeTestRule.waitForIdle()
    }

    private suspend fun resetPersistentData(container: AppContainer) {
        container.taskRepository.clearAll()
        container.habitRepository.clearAll()
        container.focusRepository.clearAll()
    }

    
    private suspend fun seedMockData(container: AppContainer) {
        val zone = ZoneId.systemDefault()
        val dayStart = LocalDate.now().atStartOfDay(zone)
        fun at(days: Long, hours: Long, minutes: Long = 0): Instant =
            dayStart.plusDays(days).plusHours(hours).plusMinutes(minutes).toInstant()

        container.settingsRepository.saveThemePref(ThemeMode.LIGHT)
        container.settingsRepository.saveDailyGoalMins(120)

        // --- Habits ---
        val habitColor = HabitColor.toArgb()
        val habits = container.habitRepository
        habits.insertHabit(
            HabitEntity(
                id = "habit_water",
                name = "Drink Water",
                frequency = HabitFrequency.DAILY,
                stackName = "Morning Routine",
                stackOrder = 1,
                createdAt = Instant.now(),
                colorValue = habitColor,
            )
        )
        habits.insertHabit(
            HabitEntity(
                id = "habit_meditate",
                name = "Meditation",
                frequency = HabitFrequency.DAILY,
                stackName = "Morning Routine",
                stackOrder = 2,
                createdAt = Instant.now(),
                colorValue = habitColor,
            )
        )
        habits.insertCompletion(
            HabitCompletionEntity(habitId = "habit_meditate", completionDate = at(0, 7, 15))
        )
        habits.insertHabit(
            HabitEntity(
                id = "habit_read",
                name = "Read 20 Pages",
                frequency = HabitFrequency.WEEKLY_EXACT,
                targetWeekdays = "[${LocalDate.now().dayOfWeek.value}]",
                notes = "Keep the momentum going before lunch.",
                createdAt = Instant.now(),
                colorValue = habitColor,
            )
        )
        habits.insertHabit(
            HabitEntity(
                id = "habit_workout",
                name = "Workout",
                frequency = HabitFrequency.WEEKLY_FLEXIBLE,
                targetDaysPerWeek = 3,
                createdAt = Instant.now(),
                colorValue = habitColor,
            )
        )
        for (daysAgo in 1L..3L) {
            habits.insertCompletion(
                HabitCompletionEntity(
                    habitId = "habit_workout",
                    completionDate = at(-daysAgo, 18)
                )
            )
        }

        // --- Tasks ---
        val tasks = container.taskRepository
        tasks.insertTask(
            TaskEntity(
                id = "task_proposal",
                title = "Finish project proposal",
                description = "Polish the scope, milestones, and budget notes.",
                dueDate = at(0, -3),
                location = "Home office",
                priority = Priority.HIGH,
                size = TaskSize.LARGE,
                category = "Work",
                createdAt = at(-2, 0),
            )
        )
        tasks.insertSubtask(
            SubtaskEntity(id = "task_proposal_1", taskId = "task_proposal", title = "Review outline")
        )
        tasks.insertSubtask(
            SubtaskEntity(
                id = "task_proposal_2",
                taskId = "task_proposal",
                title = "Check budget table"
            )
        )
        tasks.insertTask(
            TaskEntity(
                id = "task_client",
                title = "Reply to client email",
                description = "Answer the questions from yesterday's review.",
                dueDate = at(0, 23),
                location = "Inbox",
                priority = Priority.VERY_HIGH,
                size = TaskSize.SMALL,
                category = "Communication",
                createdAt = at(-1, 0),
                reminders = listOf(at(0, 9, 30)),
            )
        )
        tasks.insertTask(
            TaskEntity(
                id = "task_retro",
                title = "Prepare sprint retro",
                description = "Collect highlights and blockers for Friday.",
                dueDate = at(2, 0),
                location = "Meeting room",
                category = "Work",
                createdAt = at(0, 0),
            )
        )
        tasks.insertTask(
            TaskEntity(
                id = "task_backup",
                title = "Back up phone photos",
                description = "Archive the last trip and mark it complete.",
                dueDate = at(-1, 0),
                location = "Laptop",
                priority = Priority.LOW,
                size = TaskSize.SMALL,
                category = "Personal",
                isCompleted = true,
                completedAt = at(0, -20),
                createdAt = at(-4, 0),
            )
        )

        // --- Focus ---
        // The system modes are normally seeded lazily by FocusViewModel; the sessions below
        // reference them via a RESTRICT foreign key, so seed them here first.
        val focus = container.focusRepository
        focus.insertModeWithPhases(
            FocusModeEntity(
                id = FocusModeEntity.SYSTEM_FLEXIBLE_ID,
                name = "Flexible",
                type = FocusModeType.FLEXIBLE,
                isSystem = true
            ),
            listOf(
                SessionPhaseEntity(
                    modeId = FocusModeEntity.SYSTEM_FLEXIBLE_ID,
                    type = PhaseType.FOCUS,
                    durationMinutes = -1,
                    orderIndex = 0
                )
            ),
        )
        focus.insertModeWithPhases(
            FocusModeEntity(
                id = FocusModeEntity.SYSTEM_POMODORO_ID,
                name = "Pomodoro Classic",
                type = FocusModeType.POMODORO,
                isSystem = true
            ),
            buildList {
                repeat(4) { cycle ->
                    add(
                        SessionPhaseEntity(
                            modeId = FocusModeEntity.SYSTEM_POMODORO_ID,
                            type = PhaseType.FOCUS,
                            durationMinutes = 25,
                            orderIndex = cycle * 2
                        )
                    )
                    add(
                        SessionPhaseEntity(
                            modeId = FocusModeEntity.SYSTEM_POMODORO_ID,
                            type = PhaseType.REST,
                            durationMinutes = if (cycle == 3) 15 else 5,
                            orderIndex = cycle * 2 + 1
                        )
                    )
                }
            },
        )

        val deepWorkTagId = "tag_deep_work"
        focus.insertTag(
            FocusTagEntity(id = deepWorkTagId, name = "Deep Work", colorValue = 0xFF3F51B5.toInt())
        )

        fun pastSession(id: String, modeId: String, start: Instant, end: Instant) =
            FocusSessionEntity(
                id = id,
                modeId = modeId,
                startTime = start,
                endTime = end,
                totalSecondsFocused = (end.epochSecond - start.epochSecond).toInt(),
                isCompleted = true,
                tagId = deepWorkTagId,
                targetType = FocusTargetType.TAG,
                targetId = deepWorkTagId,
                targetLabel = "Deep Work",
            )
        focus.insertSession(
            pastSession("session_1", FocusModeEntity.SYSTEM_FLEXIBLE_ID, at(0, 8, 15), at(0, 9, 0))
        )
        focus.insertSession(
            pastSession("session_2", FocusModeEntity.SYSTEM_FLEXIBLE_ID, at(0, 13, 30), at(0, 14, 10))
        )
        focus.insertSession(
            pastSession("session_3", FocusModeEntity.SYSTEM_POMODORO_ID, at(-1, 16), at(-1, 16, 30))
        )

        // --- User profile ---
        // XP as if the seeded activity had been done in-app.
        val completedTasks = 1
        val habitCompletions = 4
        val focusMinutes = 45 + 40 + 30
        val seededXp = completedTasks * ExperienceEngine.XP_PER_TASK +
            habitCompletions * ExperienceEngine.XP_PER_HABIT +
            focusMinutes * ExperienceEngine.XP_PER_FOCUS_MINS
        container.userRepository.insertUserProfile(
            UserProfileEntity(
                name = "Bobert",
                accountCreated = Instant.now(),
                totalXp = seededXp
            )
        )
    }

    private fun takeScreenshot(name: String) {
        val bitmap = composeTestRule.onRoot().captureToImage().asAndroidBitmap()
        // AGP's additionalTestOutputDir is pulled to build/outputs/ before the app (and with it
        // any app-private directory such as getExternalFilesDir) is uninstalled and wiped.
        val outputDir = InstrumentationRegistry.getArguments().getString("additionalTestOutputDir")
        val dir = outputDir?.let { File(it) }
            ?: targetContext.getExternalFilesDir(android.os.Environment.DIRECTORY_PICTURES)
        val file = File(dir, "$name.png")
        FileOutputStream(file).use { out ->
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
        }
        println("Screenshot saved to: ${file.absolutePath}")
    }
}
